rfcs/0000-lifecycle-crd.md

```diff
@@ -0,0 +1,137 @@
+(with a label specifying the layer diff id for each os). Because of this, a cve present in the lifecycle that has been 
+patched upstream, requires any user to build the kpack specific lifecycle image themselves if the kpack provided one
+has not been bumped.
+
+As we [move to deprecate windows](https://github.com/buildpacks-community/kpack/discussions/1366), there is an opportunity for us to utilize
+the upstream lifecycle image shipped by CNB, but switching to that image without changing the interface for providing the
+lifecycle to kpack may open the door to incompatibilities if the wrong image is used on a particular kpack version.
+
+**Outcome:**
+
+Users will be able to create a CR that references a lifecycle image. The file structure in the image must match 
+that of the [CNB Lifecycle image](https://hub.docker.com/r/buildpacksio/lifecycle/tags).
+
+```
+.
+└── cnb/
+    └── lifecycle/
+        ├── analyzer
+        ├── builder
+        ├── detector
+        ├── exporter
+        ├── launcher
+        ├── lifecycle.toml
+        └── restorer
+```
+
+They will be able to use this lifecycle CR in any builder. A lifecycle will be shipped out of the box in kpack and 
+users will not be required to provide a lifecycle reference in the builder if they do not wish to override the default.
+Users should see no difference in functionality compared to the existing kpack lifecycle if they do not choose to 
+provide their own lifecycle CR.
+
+**Actions to take:**
+
+A new ClusterLifecycle CRD will be created. It will be very similar in structure to the ClusterStack Resource. 
+One open question is whether we should follow the Clusterstack model and keep this resource strictly cluster scoped or
+follow the ClusterBuildpack pattern and provide a namespaced scoped version of the CRD. 
+
+This is an example instance of the proposed CRD:
+```yaml
+apiVersion: kpack.io/v1alpha2
+kind: ClusterLifecycle
+metadata:
+  name: sample-cluster-lifecycle
+spec:
+  serviceAccountRef:
+    name: sample-sa
+    namespace: sample-namespace
+  image: buildpacksio/lifecycle@sha256:f4ce143ea6bbc6b5be5f4d151aba8214324adb93bbd7e3b1f43cd134ad450bf7
+```
+
+If we were to create a namespaced version of this CRD it would look like this. 
+Similar to Buildpack CRs, ClusterBuilders will not be able to use the namespaced CR:
+```yaml
+apiVersion: kpack.io/v1alpha2
+kind: Lifecycle
+metadata:
+  name: sample-cluster-lifecycle
+  namespace: test
+spec:
+  serviceAccountName: sample-sa
+  image: buildpacksio/lifecycle@sha256:f4ce143ea6bbc6b5be5f4d151aba8214324adb93bbd7e3b1f43cd134ad450bf7
+```
+
+The existing lifecycle reconciler can be updated to reconcile this new CRD instead of ConfigMaps.
+
+A new field will be added to the ClusterBuilder/Builder resources allowing the user to provide a reference to a lifecycle CR.
+This field will be optional and will default to the highest semantic version that is supported by the buildpacks in the builder and the platform. 
```

Comment:
Is that basically like how we hard-code the lifecycle version here? https://github.com/buildpacks-community/kpack/blob/e1722c0dbbf6d8ff3c20a84aa730a5d7e9ac221e/hack/lifecycle/main.go#L35

Would the future world deprecate that part of the code entirely? Like, if no `ClusterLifecycle` is specified do you get a default `ClusterLifecycle` or do you fallback to pulling the layer out of the `lifecycleImage`?

Reply:
ya this would get deprecated and we would pull in an upstream lifecycle in the ci pipeline instead of building our own like this file is doing. 

we would still ship a default ClusterLifecycle to ease burden on our users and avoid a breaking change on upgrade